text/0000-pin_and_move.md

```diff
@@ -0,0 +1,381 @@
+- Feature Name: pin_and_move
+- Start Date: 2018-02-19
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Introduce new APIs to libcore / libstd to serve as safe abstractions for data
+which cannot be safely moved around.
+
+# Motivation
+[motivation]: #motivation
+
+A longstanding problem for Rust has been dealing with types that should not be
+moved. A common motivation for this is when a struct contains a pointer into
+its own representation - moving that struct would invalidate that pointer. This
+use case has become especially important recently with work on generators.
+Because generators essentially reify a stackframe into an object that can be
+manipulated in code, it is likely for idiomatic usage of a generator to result
+in such a self-referential type, if it is allowed.
+
+This proposal adds an API to std which would allow you to guarantee that a
+particular value will never move again, enabling safe APIs that rely on
+self-references to exist.
+
+# Explanation
+[explanation]: #explanation
+
+## The `Move` auto trait
+
+This new auto trait is added to the `core::marker` and `std::marker` modules:
+
+```rust
+pub unsafe auto trait Move { }
+```
+
+A type implements `Move` if in its stack representation, it does not contain
+internal references to other positions within its stack representation. Nearly
+every type in Rust is `Move`.
+
+Positive impls of `Move` are added for types which contain pointers to generic
+types, but do not contain those types in their stack representation, e.g:
+
+```rust
+unsafe impl<'a, T: ?Sized> Move for &'a T { }
+unsafe impl<'a, T: ?Sized> Move for &'a mut T { }
+unsafe impl<'a, T: ?Sized> Move for Box<T> { }
+unsafe impl<"a, T> Move for Vec<T> { }
+// etc
+```
+
+This trait is a lang item, but only to generate negative impls for certain
+generators. Unlike previous `?Move` proposals, and unlike some traits like
+`Sized` and `Copy`, this trait does not impose any particular semantics on
+types that do or don't implement it.
```

Comment:
That can't literally be true, if it had no implications it would be useless. Indeed under unresolved questions the definition "it's safe to convert between `&mut T` and `Pin<T>`" is given. Is this paragraph trying to say the trait isn't "language magic" that impacts fundamental rules the compiler checks?

Reply:
yes I meant it doesn't interact with the compiler in any special way. Its just used by `Pin` to enforce invariants.